### PR TITLE
[serde-reflection] support recursive types

### PR DESCRIPTION
**Pull request by matbd**
## Summary

This PR aims to address #5 by introducing a more powerful unification algorithm with proper variables and term-sharing.

## Test Plan

included unit tests

### CODE REVIEW

File: serde-reflection/src/format.rs [hunk 1]
```diff
@@ -261,6 +321,92 @@ where
...
+        Self(Rc::new(RefCell::new(content)))
+    }
+
+    fn borrow(&self) -> Ref<Option<T>> {
+        self.0.as_ref().borrow()
+    }
+
+    fn borrow_mut(&mut self) -> RefMut<Option<T>> {
+        self.0.as_ref().borrow_mut()
+    }
```
Comment (bmwill) [inline serde-reflection/src/format.rs]: nit: maybe use the Borrow/BorrowMut traits?
(i'm fine if you do leave this as is)

File: serde-reflection/src/format.rs [hunk 2]
```diff
@@ -85,12 +99,16 @@ pub struct Named<T> {
...
+pub struct Variable<T>(Rc<RefCell<Option<T>>>);
+
 #[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 #[serde(rename_all = "UPPERCASE")]
 /// Description of a variant in an enum.
 pub enum VariantFormat {
-    /// A variant whose format is yet unknown.
-    Unknown,
+    /// (used internally for tracing) A variant whose format is initially unknown.
+    Variable(Variable<VariantFormat>),
```
Comment (bmwill) [inline serde-reflection/src/format.rs:111]: If this is for internal use only then you could always #[doc(hidden)] this

File: serde-reflection/src/de.rs [hunk 3]
```diff
@@ -290,12 +278,8 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
     where
         V: Visitor<'de>,
     {
-        self.format
-            .unify(Format::Tuple(vec![Format::Unknown; len]))?;
-        let formats = match self.format {
-            Format::Tuple(x) => x,
-            _ => unreachable!(),
-        };
+        let mut formats: Vec<_> = std::iter::repeat_with(Format::unknown).take(len).collect();
```
Comment (bmwill) [inline serde-reflection/src/de.rs:281]: Clever way to rewrite `vec![]`

File: serde-reflection/src/format.rs [hunk 4]
```diff
@@ -332,11 +341,11 @@ impl<T> Variable<T> {
         Self(Rc::new(RefCell::new(content)))
     }
 
-    fn borrow(&self) -> Ref<Option<T>> {
+    pub fn borrow(&self) -> Ref<Option<T>> {
         self.0.as_ref().borrow()
     }
 
-    fn borrow_mut(&mut self) -> RefMut<Option<T>> {
+    pub fn borrow_mut(&self) -> RefMut<Option<T>> {
```
Comment (bmwill) [inline serde-reflection/src/format.rs:348]: Why the change on mutability? and i'm expecting that you want these functions to be part of the public API?

File: serde-reflection/Cargo.toml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "serde-reflection"
-version = "0.1.0"
+version = "0.2.0"
```
Comment (bmwill) [inline serde-reflection/Cargo.toml:3]: Good this is a breaking change :)

File: serde-reflection/src/format.rs [hunk 6]
```diff
@@ -261,6 +321,92 @@ where
...
+
+    fn visit_mut(&mut self, f: &mut dyn FnMut(&mut Format) -> Result<()>) -> Result<()> {
+        match self.borrow_mut().deref_mut() {
+            None => Err(Error::UnknownFormat),
+            Some(value) => value.visit_mut(f),
+        }
+    }
+
+    fn unify(&mut self, _other: Variable<T>) -> Result<()> {
+        // TODO
```
[the remark this answers is not in the document]
Comment (matbd) [reply serde-reflection/src/format.rs]: Actually, we don't use this function (and cannot easily implement it without making `FormatHolder` bigger) so I removed the TODO in the next diff. Let me add a comment.

File: serde-reflection/src/format.rs [hunk 7]
```diff
@@ -261,6 +321,92 @@ where
...
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(_deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        use serde::de::Error;
+        Err(D::Error::custom("Cannot deserialize variables"))
+    }
+}
```
[the remark this answers is not in the document]
Comment (matbd) [reply serde-reflection/src/format.rs:389]: Let me try `deserialize_with`.

File: serde-reflection/src/format.rs (hunk 2 above)
Comment (matbd) [reply serde-reflection/src/format.rs:111]: I'm thinking a user may still wish to write `Variable(_) => unreachable!()` instead of `_ => unreachable!()` at the end of a pattern matching.

I also thought about having a separate public type but one would need to change trace_type and trace_format APIs (to not return a temporary format) and make unit tests more complicated.

File: serde-reflection/src/de.rs (hunk 3 above)
Comment (matbd) [reply serde-reflection/src/de.rs:281]: Not sure if this comment shows approval or disapproval :)
To clarify, just in case, `vec![Format::unknown(); len]` would be sharing the same variable `len` times instead of creating `len` distinct variables.

File: serde-reflection/src/format.rs (hunk 4 above)
Comment (matbd) [reply serde-reflection/src/format.rs:348]: So TBH it's public mostly for unit tests.
On second thought, I made the type more precise (although possibly less intuitive) because I found no reason no too.

File: serde-reflection/src/format.rs (hunk 1 above)
Comment (matbd) [reply serde-reflection/src/format.rs]: I tried and unfortunately (or fortunately for type safety of Rust :)), the following code doesn't work: 
```
fn borrow(&self) -> &Option<T> {  // <- & instead of Ref is required by the trait
    self.0.as_ref().borrow().deref() // but now we cannot delay the "unborrowing" of the refcell so lifetime checks fail.
}
```

File: serde-reflection/src/de.rs (hunk 3 above)
Comment (bmwill) [reply serde-reflection/src/de.rs:281]: Yeah it depends on how Clone is implemented, as it will call clone on the type provided `len` times.

File: serde-reflection/src/format.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (matbd) [reply serde-reflection/src/format.rs]: That worked.

File: serde-reflection/src/format.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (matbd) [reply serde-reflection/src/format.rs]: done

File: serde-reflection/src/de.rs (hunk 3 above)
Comment (matbd) [reply serde-reflection/src/de.rs:281]: Right. Here, we want `Clone` to create sharing.
